### ISSUE TITLE
ctypes array cannot be returned by value on ARM64

### ISSUE THREAD

**Issue by pdegroote** [python/cpython#110190]
# Bug report

### Bug description:

Dear,

I'm wrapping a C-library with the ctypes module on Darwin (ARM64). I have a function that simply initializes a C-struct and returns the struct by value. The C-struct contains one simple C-array type. When using ArrayTypes in Python, it seems the struct is not returned properly.

On Windows and Linux platforms the same code does work as expected.

The strange thing is that if I lay out the struct explicitly with singular types in Python (as I expect the layout of the struct is the same), it does work as expected on all tried platforms.

To reproduce, I first create a simple .c file with the struct and the initializing function. In Python, I give code that I use to compile (with gcc), and then three different ways to wrap the struct and function in Python using ctypes. I believe the three ways should be equivalent (and they are on Windows and Linux), but not on Darwin (ARM64).

I saved this file as ``ctypesexample.c``
```c

struct example{double coords[3];};
struct example set_defaults(){
    struct example ex;
    ex.coords[0] = 1.0;
    ex.coords[1] = 2.0;
    ex.coords[2] = 3.0;
    return ex;}
```

I saved this file as ``reproduce.py``
```python

import os
import subprocess
import ctypes

def compile():
    """Compile the c file using gcc."""
    subprocess.call(["gcc","-c","-Wall","-Werror","-fpic","ctypesexample.c"])
    subprocess.call(["gcc","-shared","-o","libctypesexample","ctypesexample.o"])
    assert os.path.isfile('libctypesexample')

# Three ways to wrap the struct:
class Example1(ctypes.Structure):
    _fields_ = [
            ('coords', ctypes.c_double * 3)
               ]

class CoordType(ctypes.Array):
    _type_ = ctypes.c_double
    _length_ = 3

class Example2(ctypes.Structure):
    _fields_ = [('coords', CoordType)]
    
class Example3(ctypes.Structure):
    _fields_ = [
            ('x', ctypes.c_double),
            ('y', ctypes.c_double),
            ('z', ctypes.c_double)
               ]

def run():
    # Load the shared library
    so_location = os.path.abspath("libctypesexample")
    _lib = ctypes.cdll.LoadLibrary(so_location)
    _lib.set_defaults.restype = Example1
    o = _lib.set_defaults()
    print(f"Example1: ({o.coords[0]}, {o.coords[1]}, {o.coords[2]})")
    
    _lib.set_defaults.restype = Example2
    o = _lib.set_defaults()
    print(f"Example2: ({o.coords[0]}, {o.coords[1]}, {o.coords[2]})")

    _lib.set_defaults.restype = Example3
    o = _lib.set_defaults()
    print(f"Example3: ({o.x}, {o.y}, {o.z})")

if __name__ == "__main__":
    if not os.path.isfile('libctypesexample'):
        compile()
    run()
```

I tried this code on Linux (debian) -- both in Docker and WSL. And this works perfectly. Equivalent code in Windows is also working fine. In the past, I have run similar code as above on many different versions of Python (3.5-3.11) on both platforms for years (and still) without problems. For example on WSL with the following platform info

`uname_result(system='Linux', release='5.15.90.1-microsoft-standard-WSL2', version='#1 SMP Fri Jan 27 02:56:13 UTC 2023', machine='x86_64')`

and Python:

Python 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0] on linux

I get this correct output:

```
Example1: (1.0, 2.0, 3.0)
Example2: (1.0, 2.0, 3.0)
Example3: (1.0, 2.0, 3.0)
```

While on **Darwin (ARM)** with the following platform info:

`system='Darwin', release='22.1.0', version='Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103', machine='arm64')`

and Python

Python 3.11.4 (main, Jun 12 2023, 11:39:45) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin

I get the following output:

```
Example1: (3.025693809e-314, 3.0256938013e-314, 2.1481083817e-314)
Example2: (3.025693809e-314, 3.0256938013e-314, 2.1481083817e-314)
Example3: (1.0, 2.0, 3.0)
```

In summary, the struct doesn't seem to be properly returned as value using array types, though with a literal listing of the types, it seems there is no problem. So there is a workaround, but as we're using ctypes for automatic wrapping through reflection and the code works properly on Windows and Linux, it would be quite some work to implement the workaround.

I hope my bug report finds you without any errors, and many thanks for giving the opportunity to report.

Best regards

### CPython versions tested on:

3.10, 3.11

### Operating systems tested on:

Linux, macOS, Windows

### Linked PRs
* gh-112604
* gh-112766
* gh-112767
* gh-112818
* gh-112829
* gh-112830
* gh-112959
* gh-113167
* gh-113170
* gh-114753
* gh-114774
* gh-114775
* gh-118233

**Comment by sunmy2019** [comment]
I looked into it. `ctypes` utilizes `libffi` to make foreign function calls, and `ctypes` generates wrong signatures.

Long words in short: `stgdict->ffi_type_pointer` is wrong. Current behavior would cause return types interpreted as `ctypes.c_void_p * 3`.

```py

class Example1(ctypes.Structure):
    _fields_ = [("coords", ctypes.c_double * 3)]


class Example2(ctypes.Structure):
    _fields_ = [("coords", ctypes.c_void_p * 3)]


# Three ways to wrap the struct:
class Example3(ctypes.Structure):
    _fields_ = [("x", ctypes.c_void_p), ("y", ctypes.c_void_p), ("z", ctypes.c_void_p)]


def run():
    # Load the shared library
    so_location = os.path.abspath("libctypesexample")
    _lib = ctypes.cdll.LoadLibrary(so_location)

    _lib.set_defaults.restype = Example1
    o = _lib.set_defaults()
    print(
        f"Example1: ({float.hex(o.coords[0])}, {float.hex(o.coords[1])}, {float.hex(o.coords[2])})"
    )

    _lib.set_defaults.restype = Example2
    o = _lib.set_defaults()
    print(f"Example2: ({o.coords[0]:x}, {o.coords[1]:x}, {o.coords[2]})")

    _lib.set_defaults.restype = Example3
    o = _lib.set_defaults()
    print(f"Example3: ({o.x:x}, {o.y:x}, {o.z})")
```
```
Example1: (0x0.0ffffcc6b11d0p-1022, 0x0.0ffff95567ef8p-1022, 0x0.0p+0)
Example2: (ffffcc6b11d0, ffff95567ef8, None)
Example3: (ffffcc6b11d0, ffff95567ef8, None)
```

Notice the binary representation of Example1 has `ffffcc6b11d0` which is same as Example2 and Example3.

**Comment by sunmy2019** [comment]
The special handling is here:
https://github.com/python/cpython/blob/21a6263020db17020b2886f996bc23aa8cb7fbdf/Modules/_ctypes/stgdict.c#L700-L710

You seem to hit one special HVA calling convention of aarch64 which is currently not handled.
https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170#return-values

> Integral values are returned in x0.
> 
> Floating-point values are returned in s0, d0, or v0, as appropriate.
> 
> A type is considered to be an HFA or HVA if all of the following hold:
> 
> It's non-empty,
> It doesn't have any non-trivial default or copy constructors, destructors, or assignment operators,
> All of its members have the same HFA or HVA type, or are float, double, or neon types that match the other members' HFA or HVA types.
> HVA values with four or fewer elements are returned in s0-s3, d0-d3, or v0-v3, as appropriate.
> 
> Types returned by value are handled differently depending on whether they have certain properties, and whether the function is a non-static member function. Types which have all of these properties,
> 
> they're aggregate by the C++14 standard definition, that is, they have no user-provided constructors, no private or protected non-static data members, no base classes, and no virtual functions, and
> they have a trivial copy-assignment operator, and
> they have a trivial destructor,
> and are returned by non-member functions or static member functions, use the following return style:
> 
> **Types that are HFAs with four or fewer elements are returned in s0-s3, d0-d3, or v0-v3, as appropriate.**
> Types less than or equal to 8 bytes are returned in x0.
> Types less than or equal to 16 bytes are returned in x0 and x1, with x0 containing the lower-order 8 bytes.
> For other aggregate types, the caller shall reserve a block of memory of sufficient size and alignment to hold the result. The address of the memory block shall be passed as an additional argument to the function in x8. The callee may modify the result memory block at any point during the execution of the subroutine. The callee isn't required to preserve the value stored in x8.
> All other types use this convention:
> 
> The caller shall reserve a block of memory of sufficient size and alignment to hold the result. The address of the memory block shall be passed as an additional argument to the function in x0, or x1 if $this is passed in x0. The callee may modify the result memory block at any point during the execution of the subroutine. The callee returns the address of the memory block in x0.

**Comment by sunmy2019** [comment]
Now the cause is clear, `ctypes` did not handle one special rule of aarch64, but fixing it is beyond my ability.

**Comment by diegorusso** [comment]
Hello, I'm looking at this issue and working on a fix.

**Comment by diegorusso** [comment]
Hello, I've investigated this further and actually this was happening across Arm platforms including Linux aarch64 and aarch32 and not just on Darwin. I've raised the PR that fixes the issue providing information and context. Feel free to review it and any feedback is welcome.

**Comment by ambv** [comment]
This is now fixed in (to be released) 3.11.8, 3.12.1, and 3.13.0 alpha 3.
Thanks, Diego! ✨ 🍰 ✨

**Comment by kulikjak** [comment]
Hi, just FYI, I am seeing similar issues on Solaris SPARC as well.

**Comment by diegorusso** [comment]
The [PowerPC ABI document](https://files.openpower.foundation/s/cfA2oFPXbbZwEBK) states: "A member of a homogeneous aggregate of multiple like data types passed in **up to eight floating-point registers**"

We put together a small example to prove this is the case: https://godbolt.org/z/fGb58Ebbs sum8 passes all values in float registers whilst sum9 passes all values using integer registers (and split the struct between integer registers and memory)
For CPython the MAX_STRUCT_SIZE for PPC64LE should be 64 bytes (8 bytes * 8 registers)

For PPC64LE I am already working on an and try to improve testing coverage as well.

**Comment by diegorusso** [comment]
@kulikjak thanks for raising the issue. Unfortunately I don't have access to Solaris SPARC machines but with my [latest PR](https://github.com/python/cpython/pull/112959) it's easier to add a case for a new architecture and having proper testing for it. 
Anyway, according to SCD specs (https://sparc.org/technical-documents/specifications/#SCD), SPARC should be OK with 16 bytes (default case). If you have a chance, you can test the PR above on Solaris SPARC and see if it works.

**Comment by kulikjak** [comment]
Hi Diego, thanks for help and sorry for the delay.

Sadly, even with the latest PR, I still see similar failures.

```
test_array_in_struct (test.test_ctypes.test_structures.StructureTestCase.test_array_in_struct) ... FAIL

======================================================================
FAIL: test_array_in_struct (test.test_ctypes.test_structures.StructureTestCase.test_array_in_struct)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/pythonmain/cpython-main/Lib/test/test_ctypes/test_structures.py", line 599, in test_array_in_struct
    self.assertEqual(result.data[i], float(i+1))
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1.110870131117437e-308 != 2.0
```

From my testing, it seems that all structs with arrays of length 4 or less do fail; 5 and up all seem fine. That said, I also believe that based on SPARC ABI specification, `MAX_STRUCT_SIZE` 16 should be correct. I still tried modifying it and nothing changed.

I guess it's more complex than that; I will try asking somebody who has much more ABI knowledge than me...

**Comment by zooba** [comment]
I'm seeing the same test fail on Windows ARM64, though not quite in the same place:

```
======================================================================
FAIL: test_array_in_struct (test.test_ctypes.test_structures.StructureTestCase.test_array_in_struct) [Test3C]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Program Files\Python313-arm64\Lib\test\test_ctypes\test_structures.py", line 670, in test_array_in_struct
    self.assertEqual(result, expected, sut.cls)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2.0 != 6.0
```

(I added `with self.subTest(sut.cls.__name__)` inside the loop to see that it's Test3C failing, so it's actually line 669 in main branch)

My build was using MSVC 19.37, which is known to have at least one ARM64 regression, so this may be another one. I haven't yet figured out how to confirm, so leaving some details here in case someone else knows before I get back to it.

**Comment by diegorusso** [comment]
@zooba thanks for reporting it. I'll try to have a look at this next week.

**Comment by diegorusso** [comment]
@zooba I have the fix, PR is coming.

**Comment by zooba** [comment]
Thanks for the quick fix for Windows! Sorry for hijacking the issue - I'll leave it open for the macOS issue.

**Comment by diegorusso** [comment]
The macOS issue has been fixed since the first PR but then the tests refactoring exposed the bug on different platforms. I mean we can keep this issue open until we cover all different platforms but I guess at some point we need to draw a line?
Should we rename/repurpose the issue for multi platforms? tier-1 and tier-2 platforms should be covered with the changes made so far.
I'm OK with whatever people think it is appropriate.

**Comment by zooba** [comment]
Let's just label it "ARM64" instead of Darwin, and if the only remaining issue is SPARC then we can probably close until someone has a machine handy to repro and test. I'll leave open for now in case someone already does, but if there's nothing heard, it can be closed.

**Comment by kulikjak** [comment]
Hi. I found some time to dig into this more and currently I suspect gcc to be the culprit - both Python and libffi seem to do what is expected. But it's definitely too soon to say for sure - I have asked those more knowledgeable internally to confirm that, so we'll see.

(It seems that gcc doesn't conform to ABI when a struct with small float/double arrays is returned from a function - when I change the test structs from one double array to several doubles, it works as expected).

I hope to find a fix soon, but I understand if you want to close this (although this seems to be a general SPARC issue, so Linux SPARC is likely affected as well). If so, can we perhaps skip those tests on SPARC?

**Comment by diegorusso** [comment]
@kulikjak thanks for the update!

**Comment by thesamesam** [comment]
Is there a link to the GCC bug? Thanks.

**Comment by kulikjak** [comment]
> Is there a link to the GCC bug? Thanks.

Sure, here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114416

**Comment by encukou** [comment]
I just merged the SPARC PR. Sorry for the delay.
As it's an unsupported/untested platform, I'd prefer if you carry backports downstream. (If that's an issue, let's talk after 3.13.0 comes out -- we definitely won't backport it in the release candidate phase.)

**Comment by kulikjak** [comment]
Thank you for the merge!

Since the sparc `define` check definitely won't break any supported platforms, I think it's safe to backport it to 3.13. That said, I completely understand that it's not something for the release candidate now (and we can certainly patch it ourselves).
